### ISSUE TITLE
click should wait for animation to finish

### ISSUE THREAD

**Issue by thesunny** [webdriverio/webdriverio#3106]
**Is your feature request related to a problem? Please describe.**
Modern interfaces use animations heavily to, for example, animate dialogs, slide sidebars in/out etc.

In order to handle this, right now, the suggestion is to write a custom command as per the issue submitted here: https://github.com/webdriverio/webdriverio/issues/429

While animated UI was rare before, now that it is common, it would be a benefit to handle this in webdriverio. This would be similar to the idea that `click` waits for the element to exist and scrolls to it before clicking. The `click` command is a smart function and is not a nieve implementation of just clicking.

**Describe the solution you'd like**
When calling `click`, the library could run a test as simple as getting the coordinates of the element, waiting until the next `requestAnimationFrame` and if the new coordinates are the same, we know that the element is not animating. If it's not animating, execute the click command. If it is animating, then we do another `requestAnimationFrame` test. There can be a timeout in the unusual case that the element animates forever.

**Describe alternatives you've considered**
This works as described in the Cypress test runner, an alternative to WebdriverIO. It works really well.

**Additional context**
n/a

**Comment by abjerstedt** [comment]
Personally, i don't agree with making click do anything other than send a click. There are cases you might actually want to catch an animation taking to long, or the app falls into a broken animation.

That said, we ourselves have custom implementations of click that we made that do what you are asking for. At most, i think we should create a secondary command that does this; but we shouldn't change the default behavior.

**Comment by thesunny** [comment]
@abjerstedt 

I can understand the desire to not over-complicate commands and I would agree that we don't want to do more than is expected of a command like click if that is your concern.

I this case, I wonder if seeing this as an over-reach might be partially influenced by the fact that the problem has already been solved for you and/or you are used to its current behavior.

Also note that the use case you mentioned of ensuring an animation doesn't take too long which I agree is useful and valuable would still be available with the existence of this feature. It seems like no code changes would even be required to keep such tests working.

Here's why I respectfully think it has more benefits than downsides for this feature to be in WebdriverIO:

- **Consistency:** It is consistent with the idea that WebdriverIO is a simplified abstraction over Webdriver itself that "just works." WebdriverIO already takes steps to ensure that things just work. For example, it checks (a) if the element exists before it tries to click it and (b) it scrolls the element into view if it isn't. Especially the second is a feature that does more than just sending a click while staying within expected behavior.
- **Better for New Users**: For a new user, the expectation is that click just works. It was not immediately obvious why my code was breaking for example and I had to implement (just like you) the animation checking code. Already, in 2/2 cases the solution had to be duplicated outside the library. I anticipate similar code has been written by many people after similar bugs were discovered.
- **No known downsides:** Perhaps there is a downside, but it feels like there are few use cases where we'd want to click an element while it is still animating. Why not just wait for it to stop? Cypress, another e2e library does that same and I haven't heard any complaints about its behavior.
- **It's not deterministic:** My code was working fine and suddenly it wasn't. This was because depending on when in the animation or whether the animation had stopped by the time the code was run, the click would work or not work. This is probably one of the more persuasive reasons, in my opinion, to wait for animation to stop. A test could pass just fine, then in a different run or different environment it could fail. The users of webdriverio are not likely as experienced as its developers and a bug that triggers/doesn't trigger at random can be hard to fix.

**Comment by abjerstedt** [comment]
I'm not saying we shouldn't do something. I'm saying it didn't be the default click. We have been trying to get as close to the metal as we can with wc3 spec behavior, and this doesn't match. 

We can create a second command to do what you are asking.

**Comment by christian-bromann** [comment]
> * (b) it scrolls the element into view if it isn't

This is not something WebdriverIO is doing. It is defined in the W3C spec and taken care of the browser driver.

> * Cypress, another e2e library does that same and I haven't heard any complaints about its behavior.

Cypress uses a different automation framework that pretty much invites you to make smart decisions for clicking. WebdriverIO is relying on WebDriver and making an additional request for an action needs to be thought through carefully.

> We can create a second command to do what you are asking.

I am not sure about this. Creating a extra command that does the same as another command but in a different way can confuse people.

With that being said, I welcome the idea of making the click command smarter. The question is, how do you want to detect an animation?

**Comment by erwinheitzman** [comment]
Another possibility is to introduce a waitForAnimation(s) instead.

**Comment by thesunny** [comment]
@abjerstedt 

Sorry, to clarify, I saw your idea of adding a second command but was still advocating for integration into the single click. I guess I'm still not understanding what the purpose of a click that doesn't wait for the element to stop animating. It feels like that would be a bug in the testing code in almost every case.

@christian-bromann 

Thanks for the clarification on scrolling into view. Didn't realize that it was handled automatically.

As for how it might be handled, this is a method on a `class` I use as a wrapper around WebdriverIO to wait for animation to stop. This method definitely won't work out of the box in WebdriverIO because there are custom methods in there that you definitely wouldn't want to use in WebdriverIO, but I think it's still valuable enough to look at.

Note: The `waitFor` method in the code just blocks until a `Promise` resolves (which is the same as waiting for an `async` method to complete). It's similar to await. I know it's not best practices and I never use anything like that in my production code but I need these tests to be written by non-developers and I couldn't get fibers to work in Jest and otherwise the test code is twice as long and harder to reason about.

Also, added some comments inline to describe what's happening to my code (not in actual code)...

```js
  click(selector) {
    const el = this.$(selector)
    // make sure element is displayed first
    waitFor(el.waitForDisplayed(2000))
    // this bit wats for element to stop moving (i.e. x/y position is same).
    // Note: I'd probably check width/height in WebdriverIO but not necessary in my use case
    waitFor(
      this.client.executeAsync(function(selector, done) {
       
        const el = document.querySelector(selector)
        if (!el)
          throw new Error(
            `Couldn't find element even though we .waitForDisplayed it`
          )
        let prevRect
        function checkFinishedAnimating() {
          const nextRect = el.getBoundingClientRect()
          // if it's not the first run (i.e. no prevRect yet) and the position is the same, anim
          // is finished so call done()
          if (
            prevRect != null &&
            prevRect.x === nextRect.x &&
            prevRect.y === nextRect.y
          ) {
            done()
          } else {
            // Otherwise, set the prevRect and wait 100ms to do another check.
            // We can play with what amount of wait works best. Probably less than 100ms.
            prevRect = nextRect
            setTimeout(checkFinishedAnimating, 100)
          }
        }
        checkFinishedAnimating()
      }, selector)
    )
    // then click
    waitFor(el.click())
    return this
  }
```

I'm using this because it's working flawlessly for my use case, but I'd probably add a timeout to the `checkFinishedAnimating` section because it's possible the element never stops animating. It would also need to be async/fiber synced like the rest of the WebdriverIO commands.

**Comment by christian-bromann** [comment]
@thesunny The problem with this check is that it will be always executed and would slow down the test for every click command you invoke. Why can't you just use the `pause` command to wait until the animation is over?

**Comment by thesunny** [comment]
@christian-bromann 

Hmm... the performance implication is something to think about.

Just to clarify, this is the code I'm using personally but, as mentioned in the post, it is just an example and I was not recommending to use the code as-is in WebdriverIO.

Here's my thoughts and ideas for improvements.

**Options Argument**

So first of all, I think `click` could benefit from an `options` argument. This would solve two issues: (1) a user can decide whether they want to take the fast click path or the safe click path without having a separate `clickWhenReady` command (2) it gives us options to customize the "is it animating" checks when the default doesn't work which I'll describe later.

**Customize WebdriverIO click options**

We could also set the click `options` in a way where user's that prefer `fast` or `safe` could do so once instead of having to pass it in each time.

*What is better? Correctness or fast defaults?*

In the options scenario, one thing we would have to decide is whether the default is a `fast` click (no animating checks) or a `safe` click.

I personally prefer the `safe` click even though it is slower because it is less likely to trip up new users. Advanced users will know what they are doing when the switch to the `fast` click. Generally, I like software that is safe but allows you to take the safety off when you've become an expert. It's kind of like MySQL having defaulted to fast before switching to safe.

That said, I think I think there are valid arguments for either side. I believe just having click `options` is a huge win.

**`setTimeout` vs `requestAnimationFrame`**

In my example, I used `setTimeout` with a 100ms wait. I've done zero testing on whether that's the optimum number. I took a guess, everything worked and executed fast for my tests, and I haven't looked back. But we should do some testing with not using a `setTimeout` and instead waiting for the next animation frame (`requestAnimationFrame`) and that would be likely the shortest time between animation frames to test.

**Animation Check Options**
It also feels valuable to be able to add custom options to handle special cases. For example, the animation might not start for 100ms. We should be able to provide something like a `startOffset`. We could also provided options for `setTimeout` in ms or `requestAnimationFrame`. This would cover all bases in the end and I can see situations where each of the options could provide value. The default animation check should be reliable for most use cases though.

**Call Reduction Alternative 1: chaining the webdriver methods**
The potential for the largest performance impact is the three calls: `waitForDisplayed`, `executeAsync` and `click`. This is actually fast in local execution (I didn't notice the performance impact) but I assume this could be slow if you are using remote options like Selenium-Grid since there are 3 round trips per click instead of one.

I don't know if this is possible, but I wonder if we can send all three instructions at once and have it return after all three are executing. Not sure how Webdriver/Selenium works under the hood so there's a good chance this is not possible. Perhaps somebody with more inside knowledge could enlighten me.

**Call Reduction Alternative 2: use a single **executeAsync****
Another alternative is that we can use a single `executeAsync` call and do all the logic ourselves in a single call. I don't love this idea because we have to duplicate the behavior of the Webdriver calls (e.g. simulating the waitForDisplayed and the way a click is delivered)

**Call Reduction Alternative 3: use a single **executeAsync** with Webdriver calls**
This would be ideal but again I don't know if it's possible. It would be nice if we could make the local webdriver calls from within the `executeAsync` code. If you know more, please post.

**Comment by christian-bromann** [comment]
@thesunny Would you mind answering the following?

> Why can't you just use the `pause` command to wait until the animation is over?

I am sure you don't deal with animated elements all the time. So why not just pause for the ones that animate?

**Comment by thesunny** [comment]
@christian-bromann 

That's a good question. There are three reasons why I prefer this:

1. The animation length can change: I will tweak the animation faster or slower quite frequently to get the right feel. Using a pause instead of an animation check means that I would have to update the tests each time I made a tweak to the animation CSS.

2. It's better for new users: Animations are common and I would say that most new users are going to be confused on why their test is failing the first time they try to click an animated element. Eventually, they'll figure it out but that first use experience is important IMO. I don't need this feature since I've already overcome that new user issue but I still believe in the value of creating a good new user experience.

3. Solving it correctly: By making it part of WebdriverIO, we have many people looking at this problem and solving it correctly. For example, my code that I posted works for me, but I can see that I might hit the use case I mentioned whereby the animation starts later. Well, then I have to go back and code it again. But if this smart clicker is part of Webdriver, then somebody would only have to solve it once then it is solved forever.

**Comment by thesunny** [comment]
Possible API:

```js
browser.click('#my-button', {
  safe: true,
  checkOffsetInMs: 200, // time until we start checking for animation
  checkFrameInMs: 100, // time to wait between frames. if `0` we use requestAnimationFrame
})
```

So if we were to default to safety mode, then we explicitly turns off safety for speed:

```js
browser.click('#my-button', { safe: false })
```

Or conversely, if we default to fast mode, then we explicitly turn on safe mode. I recommend the word `safe` by the way instead of something to do with animation so that if we add other safe checks in the future, we don't need to add another one.

```js
browser.click('#my-button', { safe: true })
```

Actually, although I advocated for safety on by default, enabling it only when required feels just fine, especially since there is, at this time, a performance penalty that we haven't been able to prove that we can get around.

**Comment by christian-bromann** [comment]
two things:

1. What type of animation are we talking here? If it is just a transition the driver should be able to still click on it
2. `safe` is imo not a right property name for it, by applying it the click waits for an animation to finish, this has nothing to do with clicking something "safely"

Anyway those kind of things might be worth discussing on the protocol level so it is implemented in the driver which can handle this much more effectively. In order to build up a case, can you explain in detail, where you are experiencing issues with interacting with animated elements?

**Comment by thesunny** [comment]
1. It can't. It fails on my tests.
2. I'm open to other suggestions. I used `safe` because without it, the `click` could fail unpredictably which feels dangerous. But like I said, any improvements on naming are welcome.

So an example is if you are opening a sidebar and it slides in. Then you want to click a link in that sidebar. That could fail.

Another example (the main problem I'm having) is that dialogs fade and slide in. I think Bootstrap does this as well which is one of, if not the most popular CSS framework.

The way that the driver appears to be working now is that it gets the button coordinates then clicks at that coordinate. But it will randomly fail if the animation has moved past that coordinate.

**Comment by emilyrohrbough** [comment]
@thesunny if your css transition has not completed before you try to click on it, it shouldn't be the click functions responsibly to know and wait for it to complete. That is an implementation detail of your test page, so the test for that page should be rewritten to acknowledge the transition by either waiting for it to complete or pausing...

If you feel strongly about creating a helper to paused clicks for animations, i.e. safeClick, you could create a wrapper command around click for your usage or even a custom service hook to check for animations before allowing click or keyboard interactions. 

IMO a click function should be just that, a click. Especially because each driver has the potential of implementing the driver click different.

**Comment by thesunny** [comment]
@emilbryggare I've already done that. There is no personal benefit for me to have this feature implemented.

You say "a click function should be just that, a click." I might actually subscribe to this logic except that it is already false in Webdriver. Webdriver (not this library but the spec) already goes beyond just being a clicker. For example, it already ensures the element is displayed before it tries to attempt the click (i.e. it waits until the element has appeared in the DOM with a sensible timeout). It also scrolls the button into view if it is not in view (i.e. it looks for the element, checks to see if it is outside the viewable error, and executes a scroll command to put the element into view if possible).

I don't believe it used to do all this as I remember I had code previously that would ensure that the element was visible in the UI before it tried to click.

This suggestion feels like it adds to that natural progression of working the way people expect.

**Comment by emilyrohrbough** [comment]
> Webdriver (not this library but the spec)

Is that consistently implemented across drivers?

**Comment by christian-bromann** [comment]
> already goes beyond just being a clicker.

@thesunny you are right, which is why I was about to make a proposal to the W3C spec but therefor it would be good to get more insights into the problem.

**Comment by abjerstedt** [comment]
I STILL don't think we should put this behavior in the native click command.  

@thesunny @emilbryggare @christian-bromann Trying to reference the scroll to on the click as an example of making the click safer is kind of a strawman argument. It is IMPOSSIBLE for a user to actually click on an element if it is offscreen, unless they use JS or some other event to call it.

**Comment by emilyrohrbough** [comment]
Total side note, but @abjerstedt & @thesunny you have both tagged the wrong Emily on your comments 😋

**Comment by bjankord** [comment]
My preference would be to not overload `click` more so than it is already. I agree with @emilyrohrbough that animations are an implementation detail of your test page, so the test for that page should be rewritten to acknowledge the transition by either waiting for it to complete or pausing.

One issue I'm curious about @thesunny, it sounds like the animation check would only check postion/offset on the element clicked. Does it not account for animated elements within the element receiving the click event. e.g. You click on a button and a SVG spins inside it. The offset check doesn't seem like it would detect that animation, or if the button is animated in color.

**Comment by abjerstedt** [comment]
@thesunny I don't think anyone is arguing against your safeClick() - At least in my case, i just don't want the base method to do it. Again, I'm all for adding a different method.

**Comment by thesunny** [comment]
@abjerstedt Your position is clear and thank you for the clarity. At the same time it's not so clear (at least to me) how others feel about a safe click feature in this library.

I guess an important question which I'm not sure the answer to is who (singular or plural) gets to decide what gets merged?

I'm willing to contribute code at some point in the future but wouldn't want to dedicate time unless I was sure it would be merged assuming the code was of acceptable quality of course.

**Comment by abjerstedt** [comment]
@thesunny At the moment only Christian and I have merge rights afaik. In terms of decision making, its supposed to be the Technical committee (which again currently is the two of us).

That being said, this project is Christian's baby... and whatever he says goes :).

**Comment by BorisOsipov** [comment]
>At the same time it's not so clear (at least to me) how others feel about a safe click feature in this library.

My two cents
1) Why only click? What we should do with other action methods?
2) There should be not safe and unsafe actions. For what we need unsafe actions? to make test more flacky?

For me it is ok that currently wdio provides command implementations based on selenium spec and handle some base errors. It is up to you how handle animations\elements overlapping in your tests.

**Comment by abjerstedt** [comment]
@BorisOsipov What base errors in click are you thinking that we should handle?

**Comment by BorisOsipov** [comment]
@abjerstedt In fact we can handle and retry almost all error(exclude selenium connection\timeout errors). 
But if we want to be more conservative we can handle and retry clicks several times when we get errors like "Element not clickable at point"\"Element not interactable" - in the most cases we get such errors when element is not displayed yet, but will display in next seconds(by animations, transitions, etc)

**Comment by BorisOsipov** [comment]
@abjerstedt actually we success use such approach for all actions(retry actions several times until success) in well known java tests automation library https://github.com/selenide/selenide

**Comment by BorisOsipov** [comment]
>We currently have a retry process baked into the request handler that will attempt each request 3x in the event of an error; and a stale element handler that will refetch the entire element chain whenever you try to interact with a stale element.

That's nice. I think it is enough for now. When we will get some feedback we will be able adjust this solution somehow.

>I do not like the idea of overriding base element interactions in ways that force all of the webdriverio users to use a non webdriver spec compliant version. I'm fine with wrapping, but not overriding.

I agree.

Also I have question @abjerstedt
>request handler that will attempt each request 3x

does request handler have some timeout between attemtps?

**Comment by thesunny** [comment]
@abjerstedt it's hard to explain exactly why, and something I think I could communicate better in person, but my instinct is that creating a separate wrapper means over time it will be less and less community supported. Something like: Because it's not in wdio proper, it's a second class citizen, meaning it doesn't get first class support, which means it doesn't get used as much, which means it gets less support and over time this cycles repeats and we just end up deprecating it. While putting it in wdio proper means it gets first class support because everybody is using it.

If we are going to wrap it, probably just as fine creating a completely separate library and then at least we can justify locking the version of wdio which will prevent updates to wdio from causing bugs in the wrapper. But I personally don't have the time to keep a separate package up to date myself.

I'm going to walk back an earlier statement about `safe` option though. I think given the premise that we want the methods to be pure abstractions over `webdriver` by default which I think is a fair comment given that all the other magic actually happens in webdriver (the interface) and not wdio, perhaps a good compromise is to enable additional features in the options on a feature by feature basis. For example, `.click({waitForAnimation: true})`. We can then add other features if need be as well. This is a bad example, but things like `.click({waitForOpaque: true})` to wait until an element is `opacity === 100` for example.

I like `options` better than a completely separate method because over time, it's possible that `webdriver` (the interface) will support in its spec some of these features. Then, for example, `waitForAnimation: true` can be a `noop` and the user of the library has to make no code changes to get the same functionality. That said, the semantics doesn't matter enough to me to prevent the implementation of the feature if you prefer using something like `.smartClick({waitForAnimation: true})`. It's just a little more typing when refactoring.

**Comment by abjerstedt** [comment]
@thesunny Thats not necessarily true anymore. In version 5 - many packages are being imported into the wdio monorepo and are now maintained by the core wdio team.

Anyways - I talked to christian more about this and as long as you aren't overriding the default behavior, its probably fine.

My question next becomes, does this belong in click specifically, or should we have a wait method that returns the element so that you can chain on it? IE: PageObject.element.waitUntilElement(blah).click();?

This way the behavior can be used for all interactions.

**Comment by thesunny** [comment]
Hmm... that's an interesting proposal...

I like the simplicity of being able to add this to a `click` option or a `smartClick` option because it's easy to understand but I also like the flexibility of being able to add it anywhere. I suppose another possibility is to do both. The `option` could just be an alias for chaining before click...

**Comment by filipemir** [comment]
Just adding an opinion to this issue since the discussion seems to no longer be active: having the ability to detect whether an element transition/animated has finished (with a specifiable timeout) would be very useful in all sorts of cases. so +1

**Comment by abjerstedt** [comment]
@filipemir I think we all agree that it would be helpful ;). The question is how to implement this.

**Comment by DustinRoundy** [comment]
I agree this would be helpful in many cases. I don't necessarily agree with putting it in the click function but maybe a new .waitForAnimation() function. I think this would have many use cases!

**Comment by anto-ac** [comment]
I’m sure I’m not saying anything groundbreaking here, but this is my usual approach:

```
import { Element } from '@wdio/sync';

export function waitUntilElementHasStoppedMoving(element: Element): void {
  browser.waitUntil(() => {
    const oldX = element.getLocation('x');
    const oldY = element.getLocation('y');
    browser.pause(500);
    const newX = element.getLocation('x');
    const newY = element.getLocation('y');
    return oldX === newX && oldY === newY;
  }, 3000, `Element ${element} did not stop moving`);
}
```

Obviously this can added as a custom command. Whether it’s something you’d want on the wdio api, that’s a different story.

**Comment by drewctate** [comment]
Similar story @anto-ac 
```js
/**
 * Waits for an element to stop moving.
 * (can be used to wait for an animation to finish)
 * @param {number} timeout
 * @param {string} timeoutMsg
 * @param {number} interval
 */
browser.addCommand('waitForStopped', async function (timeout, timeoutMsg, interval) {
  this.waitForDisplayed();

  let lastCoords = await getCoordsForElement(this.elementId);
  const stoppedMoving = async () => {
    const coords = await getCoordsForElement(this.elementId);
    const notChanged = coords[0] === lastCoords[0] && coords[1] === lastCoords[1];
    console.log(lastCoords, coords);
    lastCoords = coords;
    return notChanged;
  };

  return browser.waitUntil(stoppedMoving, timeout, timeoutMsg, interval);
}, true);
```

**Comment by eseakin** [comment]
Hi, just wanted to chime in about the waitForAnimation idea. I work at a huge company and have zero control over components. That's an entire division, and they decide what animations to use or when to change them. I have to build automated integration tests for my team's tiny little piece (which requires running through several other teams' pieces, which I also have no control over), and often I don't know what the animation times are (I can certainly dig in and find them, but it's a lot of time and they might change without warning). waitForVisible doesn't work when components are still animating (we often disable buttons until the animation is done), so I've been using browser.pause() a lot -- except I don't know how long to pause for so I just put 1000. This is obviously inefficient, but it's better than all the tests failing a month from now because someone changed an animation.

It would be a great time saver to have a function that hooks into animation events and waits for them to be over before proceeding. I'm completely new to Selenium/Webdriver so maybe I just don't know how to do it. We are already looking at ~3 minutes to run through the whole product to my team's portion, so any time saved would be great. I'm going to give these waitUntilElementHasStopped functions a try. Thanks!

**Comment by anto-ac** [comment]
@eseakin See my message and the message underneath to get an idea of how you can explicitly wait for certain types or animations to complete.

**Comment by eseakin** [comment]
Yeah I just implemented a JS version of your animation watcher and it works! I still feel like it should be included with webdriverio — animations are a very standard thing on any kind of professional web app. Things like fade won’t be covered by your function and it should be possible to hook into JS animation events to know exactly when an animation ends instead of polling for it. 

Thanks for the help :)

**Comment by rubenlg** [comment]
Rather than making tests slower by waiting for animations, wouldn't it be better to "disable" CSS animations on tests? Tests are not human, they don't care if the animation is a beautiful 300ms ease-in-out or it completes in 1ms. As long as animation events keep triggering, it should be good.

Also, we don't need webdriverio to provide extra support for that, you can inject a CSS script that selectively disables some animations, or all of them. Although having some `browser.makeAllAnimationsImmediate()` helper would be super nice.

The only caveat is Shadow DOM, but I'm not sure if webdriverio supports that at all.

**Comment by janmatura** [comment]
Thank you  @thesunny for inspiration. I tried to make it easier little bit. Using separately before click() command if needed. It seem to be ultimate weapon for us using in animated SPA. Thank you for feedback. 

`browser.addCommand('waitForAnimation', function () {
        this.waitForExist(15000);
        var animating = true;

                var preLoc = this.getLocation();

                while(animating){
                    browser.pause(150);
                    postLoc = this.getLocation();
                
                    if( preLoc.x == postLoc.x &&
                        preLoc.y == postLoc.y )
                        {
                        animating = false;
                        } 
                    var spread = [postLoc.x - preLoc.x, postLoc.y - preLoc.y]
                    console.log(`WaitForAnimation: Spread x: ${spread[0]}; Spread y: ${spread[1]}`)

                    preLoc = postLoc;
               }
               console.log('WaitForAnimation: Done')
               return true;
                    
            }, true);
        }`

**Comment by christian-bromann** [comment]
@janmatura this seems like a nice cross environment solution that could be added to the click command. Here is what I would suggest:

- add a click option like `waitForAnimationTimeout` which defaults to `0`
- if set we wait for animation as shown in the example above until the timeout is reached and the continue

Any contributions on this would be welcome!

**Comment by erwinheitzman** [comment]
Hi all, I would like to resolve this issue by introducing a “stable” state which can be implemented by using the execute command and using the requestAnimationFrame function to compare the boundingbox positions and if they are the same, the element is considered stable. I think this is a nice solution as it only costs a single command execution. What do you think?

**Comment by christian-bromann** [comment]
Makes sense to me 👍

**Comment by dgp1130** [comment]
As mentioned by a few others, I feel like this is a broader problem than just waiting for animations after a click event. Any event can trigger an animation (typing into a form field, tabbing between elements, scroll-driven animations, etc.). Animations also take time and slow down tests as well as generally lead to increased flakiness, especially whenever screenshots are involved.

I want to suggest an alternative way of globally disabling animations: `prefers-reduced-motion`. A common way of handling `prefers-reduced-motion: reduced` is to disable animations in CSS. Tests can take advantage of this if we can just configure the browser to set this value.

I filed https://github.com/webdriverio/webdriverio/issues/11683 with more info but I think this can be a reasonable general solution to the animation problem in a few ways:
1. It is not tied to any specific interaction. All animations are affected, not just ones triggered by clicks.
2. Even if your site doesn't support `prefers-reduced-motion` today, it probably should. So if you need to fix a test by adding a media query on `prefers-reduced-motion`, your not _just_ fixing your tests, but also improving your overall a11y and user experience.
3. This doesn't require the test author to manually opt-in each interaction to "skip animations". This is implicitly done for all interactions.

I think there are two notable caveats to keep in mind however.
1. `prefers-reduced-motion: reduced` doesn't exactly map its semantics to "disable animations". There may be cases where a reduced motion user may want to see an animation, yet a test would benefit from disabling that animation. I'm not an a11y expert, but I'd be interested to hear what those use cases are.
2. Some animations are critical for functionality. For example, a game where a bullet moves from from a source to a target. This motion is critical to the experience and shouldn't be turned off, even for reduced motion users. I'm less concerned about this case because I think:
    a. Motion like this probably isn't using CSS animations.
    b. Functionality-critical movement is likely important to tests and _shouldn't_ be turned off in the first place. The fact that `prefers-reduced-motion: reduced` does _not_ disable this movement is a feature in my mind.

This is just one approach to consider how the animation problem can be solved more generally at the browser level.

**Comment by erwinheitzman** [comment]
We will be able to disable animations with Webdriver Bidi by injecting some JavaScript. Doing this however feels like a separate task and we can tackle both independently.

That said you can already disable animations yourself if you want by injecting some css on every page visit.

**Comment by ios-simulator1** [comment]
> I'm not saying we shouldn't do something. I'm saying it shouldn't be the default click. We have been trying to get as close to the metal as we can with wc3 spec behavior, and this doesn't match.
> 
> We can create a second command to do what you are asking.





https://laravel-storage.github.io/HlKN5rqlezbB8Jd6nouaXVs9l7MwiQQ/
